// beautiful-react-hooks/index.js
/* beautiful-react-hooks version: 0.31.1 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var useDidMount = require('./useDidMount.js');
var useWillUnmount = require('./useWillUnmount.js');
var useLifecycle = require('./useLifecycle.js');
var useWindowResize = require('./useWindowResize.js');
var useWindowScroll = require('./useWindowScroll.js');
var useDebouncedFn = require('./useDebouncedFn.js');
var useThrottledFn = require('./useThrottledFn.js');
var useMouse = require('./useMouse.js');
var useMouseEvents = require('./useMouseEvents.js');
var useMouseState = require('./useMouseState.js');
var useTimeout = require('./useTimeout.js');
var useConditionalTimeout = require('./useConditionalTimeout.js');
var useInterval = require('./useInterval.js');
var useGlobalEvent = require('./useGlobalEvent.js');
var usePreviousValue = require('./usePreviousValue.js');
var useGeolocation = require('./useGeolocation.js');
var useGeolocationEvents = require('./useGeolocationEvents.js');
var useGeolocationState = require('./useGeolocationState.js');
var useMediaQuery = require('./useMediaQuery.js');
var useValueHistory = require('./useValueHistory.js');
var useOnlineState = require('./useOnlineState.js');
var useViewportSpy = require('./useViewportSpy.js');
var useValidatedState = require('./useValidatedState.js');
var useDragEvents = require('./useDragEvents.js');
var useDrag = require('./useDrag.js');
var useDropZone = require('./useDropZone.js');
var useRequestAnimationFrame = require('./useRequestAnimationFrame.js');
var useLocalStorage = require('./useLocalStorage.js');
var useSessionStorage = require('./useSessionStorage.js');
var useStorage = require('./useStorage.js');
var useResizeObserver = require('./useResizeObserver.js');
var useDefaultedState = require('./useDefaultedState.js');
var useObservable = require('./useObservable.js');
var useSpeechSynthesis = require('./useSpeechSynthesis.js');
var useSystemVoices = require('./useSystemVoices.js');
var useRenderInfo = require('./useRenderInfo.js');



exports.useDidMount = useDidMount['default'];
exports.useWillUnmount = useWillUnmount['default'];
exports.useLifecycle = useLifecycle['default'];
exports.useWindowResize = useWindowResize['default'];
exports.useWindowScroll = useWindowScroll['default'];
exports.useDebouncedFn = useDebouncedFn['default'];
exports.useThrottledFn = useThrottledFn['default'];
exports.useMouse = useMouse['default'];
exports.useMouseEvents = useMouseEvents['default'];
exports.useMouseState = useMouseState['default'];
exports.useTimeout = useTimeout['default'];
exports.useConditionalTimeout = useConditionalTimeout['default'];
exports.useInterval = useInterval['default'];
exports.useGlobalEvent = useGlobalEvent['default'];
exports.usePreviousValue = usePreviousValue['default'];
exports.useGeolocation = useGeolocation['default'];
exports.useGeolocationEvents = useGeolocationEvents['default'];
exports.useGeolocationState = useGeolocationState['default'];
exports.useMediaQuery = useMediaQuery['default'];
exports.useValueHistory = useValueHistory['default'];
exports.useOnlineState = useOnlineState['default'];
exports.useViewportSpy = useViewportSpy['default'];
exports.useValidatedState = useValidatedState['default'];
exports.useDragEvents = useDragEvents['default'];
exports.useDrag = useDrag['default'];
exports.useDropZone = useDropZone['default'];
exports.useRequestAnimationFrame = useRequestAnimationFrame['default'];
exports.useLocalStorage = useLocalStorage['default'];
exports.useSessionStorage = useSessionStorage['default'];
exports.useStorage = useStorage['default'];
exports.useResizeObserver = useResizeObserver['default'];
exports.useDefaultedState = useDefaultedState['default'];
exports.useObservable = useObservable['default'];
exports.useSpeechSynthesis = useSpeechSynthesis['default'];
exports.useSystemVoices = useSystemVoices['default'];
exports.useRenderInfo = useRenderInfo['default'];
