// beautiful-react-hooks/useSpeechSynthesis.js
/* beautiful-react-hooks version: 0.31.1 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('./_virtual/_rollupPluginBabelHelpers.js');
var react = require('react');

var defaultOptions = {
  rate: 0,
  pitch: 0,
  volume: 1
};

var useSpeechSynthesis = function useSpeechSynthesis(text) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : defaultOptions;
  var utter = react.useMemo(function () {
    return new SpeechSynthesisUtterance(text);
  }, [text]);

  var voiceOptions = _rollupPluginBabelHelpers.objectSpread2(_rollupPluginBabelHelpers.objectSpread2({}, defaultOptions), options);

  utter.voice = voiceOptions.voice;
  react.useEffect(function () {
    utter.pitch = voiceOptions.pitch;
  }, [voiceOptions.pitch]);
  react.useEffect(function () {
    utter.rate = voiceOptions.rate;
  }, [voiceOptions.rate]);
  react.useEffect(function () {
    utter.volume = voiceOptions.volume;
  }, [voiceOptions.volume]);
  var speak = react.useCallback(function () {
    return speechSynthesis.speak(utter);
  }, [text, voiceOptions.pitch, voiceOptions.rate, voiceOptions.voice, voiceOptions.volume]);
  return {
    speak: speak,
    speechSynthUtterance: utter
  };
};

exports.default = useSpeechSynthesis;
