// beautiful-react-ui/dist/components/Forms/FileUploader/utils/useFileUploaderHandlers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = require("react");

var _beautifulReactHooks = require("beautiful-react-hooks");

var _getNextUploadsBy = _interopRequireDefault(require("./getNextUploadsBy"));

var _removeUploadAt = _interopRequireDefault(require("./removeUploadAt"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var useFileUploadHandlers = function useFileUploadHandlers(uploads, multiple, onChange) {
  var dropZoneRef = (0, _react.useRef)();
  var inputRef = (0, _react.useRef)();

  var _useDropZone = (0, _beautifulReactHooks.useDropZone)(dropZoneRef),
      onDrop = _useDropZone.onDrop,
      isOver = _useDropZone.isOver;

  var onFilesChange = (0, _react.useCallback)(function (event) {
    if (onChange && event.target.files.length > 0) {
      var getNextUploads = (0, _getNextUploadsBy.default)(multiple)(uploads);
      var nextUploads = getNextUploads(event.target.files);
      onChange(event, nextUploads);
    }
  }, [uploads, onChange]);
  var onFileRemove = (0, _react.useCallback)(function (index) {
    return function (event) {
      if (onChange) {
        var removeUpload = (0, _removeUploadAt.default)(index);
        var nextUploads = removeUpload(uploads);
        onChange(event, nextUploads);
      }
    };
  }, [uploads, onChange]);
  onDrop(function (event) {
    var getNextUploads = (0, _getNextUploadsBy.default)(multiple)(uploads);
    var nextUploads = getNextUploads(event.dataTransfer.files);
    onChange(event, nextUploads);
    event.preventDefault();
  });
  var clickHandler = (0, _react.useCallback)(function (e) {
    if (inputRef.current) {
      e.stopPropagation();
      inputRef.current.click();
    }
  }, [inputRef.current]);
  return {
    onFilesChange: onFilesChange,
    onFileRemove: onFileRemove,
    clickHandler: clickHandler,
    isOver: isOver,
    dropZoneRef: dropZoneRef,
    inputRef: inputRef
  };
};

var _default = useFileUploadHandlers;
exports.default = _default;
