// beautiful-react-ui/dist/components/Layout/NotificationsStack/NotificationRightRender.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Button = _interopRequireDefault(require("../../Elements/Button/Button"));

var _CloseIcon = _interopRequireDefault(require("../../Elements/_CloseIcon/CloseIcon"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var NotificationRightRender = function NotificationRightRender(_ref) {
  var onToggle = _ref.onToggle;
  return _react.default.createElement("div", {
    className: "notification-right-side"
  }, _react.default.createElement(_Button.default, {
    color: "transparent",
    onClick: onToggle,
    size: "small"
  }, _react.default.createElement(_CloseIcon.default, null)));
};

NotificationRightRender.propTypes = {
  onToggle: _propTypes.default.func.isRequired
};
var _default = NotificationRightRender;
exports.default = _default;
