// beautiful-react-ui/dist/components/Layout/NotificationsStack/NotificationsStack.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("core-js/modules/es6.object.assign");

require("core-js/modules/es6.symbol");

require("core-js/modules/web.dom.iterable");

require("core-js/modules/es6.array.iterator");

require("core-js/modules/es6.object.to-string");

require("core-js/modules/es6.object.keys");

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _isEqual = _interopRequireDefault(require("lodash/isEqual"));

var _Portal = _interopRequireDefault(require("../../Elements/_Portal"));

var _Notification = _interopRequireDefault(require("./Notification"));

var _IconProp = _interopRequireDefault(require("../../../shared/types/IconProp"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var NotificationsStack = function NotificationsStack(props) {
  var _classNames;

  var notifications = props.notifications,
      onChange = props.onChange,
      NotificationRender = props.NotificationRender,
      position = props.position,
      animation = props.animation,
      color = props.color,
      rest = _objectWithoutProperties(props, ["notifications", "onChange", "NotificationRender", "position", "animation", "color"]);

  var _onClose = function onClose(notification) {
    var nextNotifications = notifications.filter(function (item) {
      return !(0, _isEqual.default)(notification, item);
    });
    onChange(nextNotifications);
  };

  var classList = (0, _classnames.default)('bi-notifications-wrapper', (_classNames = {}, _defineProperty(_classNames, "bi-ns-".concat(position), !!position), _defineProperty(_classNames, "bi-ns-".concat(animation), !!animation), _defineProperty(_classNames, "bi-ns-".concat(color, "-color"), !!color), _classNames));
  return _react.default.createElement(_Portal.default, {
    id: "bi-notification-stack"
  }, _react.default.createElement("div", _extends({
    className: classList
  }, rest), notifications && notifications.map(function (item) {
    return _react.default.createElement(NotificationRender, _extends({}, item, {
      onClose: function onClose() {
        return _onClose(item);
      },
      key: item.id
    }));
  })));
};

NotificationsStack.propTypes = {
  notifications: _propTypes.default.arrayOf(_propTypes.default.shape({
    id: _propTypes.default.number.isRequired,
    content: _propTypes.default.string.isRequired,
    title: _propTypes.default.string,
    icon: _IconProp.default,
    avatar: _propTypes.default.string,
    timeout: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.number]),
    closable: _propTypes.default.bool
  })),
  onChange: _propTypes.default.func.isRequired,
  position: _propTypes.default.oneOf(['top-center', 'bottom-center', 'top-right', 'top-left', 'bottom-right', 'bottom-left']),
  NotificationRender: _propTypes.default.elementType,
  animation: _propTypes.default.oneOf(['none', 'fade', 'zoom', 'slide-right', 'slide-top', 'slide-left', 'slide-bottom']),
  color: _propTypes.default.oneOf(['info', 'success', 'warning', 'danger', 'default'])
};
NotificationsStack.defaultProps = {
  notifications: undefined,
  position: 'top-right',
  NotificationRender: _Notification.default,
  animation: 'slide-right',
  color: 'default'
};

var _default = _react.default.memo(NotificationsStack);

exports.default = _default;
