// beautiful-react-ui/dist/components/Layout/NotificationsStack/NotificationLeftRender.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Icon = _interopRequireDefault(require("../../Elements/Icon"));

var _Avatar = _interopRequireDefault(require("../../Elements/Avatar"));

var _shared = require("../../../shared");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var NotificationLeftRender = function NotificationLeftRender(_ref) {
  var icon = _ref.icon,
      avatar = _ref.avatar;
  return _react.default.createElement("div", {
    className: "notification-left-side"
  }, icon && _react.default.createElement(_Icon.default, {
    name: icon,
    size: "large"
  }), avatar && _react.default.createElement(_Avatar.default, {
    src: avatar
  }));
};

NotificationLeftRender.propTypes = {
  icon: _shared.IconProp,
  avatar: _propTypes.default.string
};
NotificationLeftRender.defaultProps = {
  icon: undefined,
  avatar: undefined
};
var _default = NotificationLeftRender;
exports.default = _default;
