// beautiful-react-ui/dist/shared/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Color", {
  enumerable: true,
  get: function get() {
    return _Color.default;
  }
});
Object.defineProperty(exports, "Size", {
  enumerable: true,
  get: function get() {
    return _Size.default;
  }
});
Object.defineProperty(exports, "Placement", {
  enumerable: true,
  get: function get() {
    return _Placement.default;
  }
});
Object.defineProperty(exports, "IconProp", {
  enumerable: true,
  get: function get() {
    return _IconProp.default;
  }
});
Object.defineProperty(exports, "PillProp", {
  enumerable: true,
  get: function get() {
    return _PillProp.default;
  }
});
Object.defineProperty(exports, "makeCallback", {
  enumerable: true,
  get: function get() {
    return _makeCallback.default;
  }
});
Object.defineProperty(exports, "makeKeyboardCallback", {
  enumerable: true,
  get: function get() {
    return _makeKeyboardCallback.default;
  }
});
Object.defineProperty(exports, "makeIconFromProp", {
  enumerable: true,
  get: function get() {
    return _makeIconFromProp.default;
  }
});
Object.defineProperty(exports, "makeSpinnerFromProp", {
  enumerable: true,
  get: function get() {
    return _makeSpinnerFromProp.default;
  }
});
Object.defineProperty(exports, "makePillFromProp", {
  enumerable: true,
  get: function get() {
    return _makePillFromProp.default;
  }
});
Object.defineProperty(exports, "emptyChildren", {
  enumerable: true,
  get: function get() {
    return _emptyChildren.default;
  }
});
Object.defineProperty(exports, "getLastChild", {
  enumerable: true,
  get: function get() {
    return _getLastChild.default;
  }
});
Object.defineProperty(exports, "getFirstChild", {
  enumerable: true,
  get: function get() {
    return _getFirstChild.default;
  }
});
Object.defineProperty(exports, "warn", {
  enumerable: true,
  get: function get() {
    return _warn.default;
  }
});
Object.defineProperty(exports, "checkOnAllowedChildren", {
  enumerable: true,
  get: function get() {
    return _checkOnAllowedChildren.default;
  }
});

var _Color = _interopRequireDefault(require("./types/Color"));

var _Size = _interopRequireDefault(require("./types/Size"));

var _Placement = _interopRequireDefault(require("./types/Placement"));

var _IconProp = _interopRequireDefault(require("./types/IconProp"));

var _PillProp = _interopRequireDefault(require("./types/PillProp"));

var _makeCallback = _interopRequireDefault(require("./functions/makeCallback"));

var _makeKeyboardCallback = _interopRequireDefault(require("./functions/makeKeyboardCallback"));

var _makeIconFromProp = _interopRequireDefault(require("./functions/makeIconFromProp"));

var _makeSpinnerFromProp = _interopRequireDefault(require("./functions/makeSpinnerFromProp"));

var _makePillFromProp = _interopRequireDefault(require("./functions/makePillFromProp"));

var _emptyChildren = _interopRequireDefault(require("./functions/emptyChildren"));

var _getLastChild = _interopRequireDefault(require("./functions/getLastChild"));

var _getFirstChild = _interopRequireDefault(require("./functions/getFirstChild"));

var _warn = _interopRequireDefault(require("./functions/warn"));

var _checkOnAllowedChildren = _interopRequireDefault(require("./functions/checkOnAllowedChildren"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
